### PR TITLE
hostname: Refactor code

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

A lot of code was repeated in different strategies.
Refactor the code to re-use existing code.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/hostname_refactor.yml
lib/ansible/modules/hostname.py

### CODE REVIEW

File: lib/ansible/modules/hostname.py [hunk 1]
```diff
@@ -459,42 +403,14 @@ def set_permanent_hostname(self, name):
             f.close()
 
 
-class OpenBSDStrategy(GenericStrategy):
+class OpenBSDStrategy(DebianStrategy):
```
Comment (bcoca) [inline lib/ansible/modules/hostname.py]: considering you are applying it to BSD .. i would move the code from Debian strategy to something like posix|gnu|oldunixy|presystemd/strategy and reasing to all those with the common approach

Comment (bcoca) [reply lib/ansible/modules/hostname.py]: HostnameCLIStrategy?

**Comment by Akasurde** [comment]
@samdoran @relrod While we are at hostname module, can you please review this as well?

**Comment by relrod** [comment]
I'll have to take a closer look. I have a local patch for removing the destructiveness of the `get_*` methods (i.e. the behavior that leads to creating a file even in check_mode), but it looks like this patch does that already, at least for some of the strategies. Will probably have to combine them together somehow.

File: lib/ansible/modules/hostname.py [hunk 2]
```diff
@@ -215,62 +219,42 @@ def get_permanent_hostname(self):
...
-            f = open(self.HOSTNAME_FILE)
-            try:
-                return f.read().strip()
-            finally:
-                f.close()
-        except Exception as e:
-            self.module.fail_json(msg="failed to read hostname: %s" %
-                                      to_native(e), exception=traceback.format_exc())
+
+        return get_file_content(self.HOSTNAME_FILE)
```
Comment (samdoran) [inline lib/ansible/modules/hostname.py:223]:
```suggestion
        return get_file_content(self.HOSTNAME_FILE, strip=True)
```

**Review by samdoran — COMMENTED** [review]
I think this is a nice consolidation overall. There are a few places where I'm worried there might be subtle differences between the line stripping behavior between what was done in each method vs how it is returned from `get_file_lines()`.

The only thing I feel we're losing with the change to `get_file_[lines|content]()` is the `Exception` is ignored in `get_file_content()` whereas currently it results in a `fail_json()` call. I think there we will catch and handle errors later when trying to write to the file, but it's a change in behavior. It'd be nice if `get_file_content()` didn't just silently pass all exceptions and that behavior was configurable.

File: changelogs/fragments/hostname_refactor.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- hostname - refactor code for DebianStrategy and GenericStrategy.
```
Comment (relrod) [inline changelogs/fragments/hostname_refactor.yml:2]: This seems outdated, this is touching more than those :)

File: lib/ansible/modules/hostname.py [hunk 4]
```diff
@@ -539,41 +411,27 @@ def get_permanent_hostname(self):
             except IOError as e:
```
Comment (relrod) [inline lib/ansible/modules/hostname.py:411]: This still does a write, right above this line. This means we're modifying the system in this function (and this function gets called in check_mode, see #66432).

But fixing this leads down a rabbit hole of other issues:
* We don't currently standardize on "UNKNOWN" vs "" when the file doesn't exist.
  * Some modules write it (again - breaking check_mode's promise) as "" and then return that as the "before" value. 
  * Others default to "UNKNOWN" if it doesn't exist.
  * FreeBSDStrategy defaults to "temporarystub" if it doesn't exist.

My proposal (#70828) is to unify them all to `""` - but this is a breaking change if anyone depends on the "before" value.

**Review by relrod — CHANGES_REQUESTED** [review]
I think this is an awesome refactor. I honestly forgot about this PR and did a similar one (#70828) but it (mine) gives up backwards compat in the name of consistency (see my inline comment below for some details). It does keep the exceptions around that @samdoran mentioned though.

This still modifies the system in check_mode in some situations, so I think we'd need to fix that in any case. (Mine stops doing that, but returns "" instead of "temporarystub" for FreeBSD, again breaking backwards compat.)

**Comment by Akasurde** [comment]
superseded by https://github.com/ansible/ansible/pull/74797

**Pull request closed without merging**